ext/liquid_c/vm.c

```diff
@@ -113,9 +114,36 @@ static void write_fixnum(VALUE output, VALUE fixnum)
     snprintf(RSTRING_PTR(output) + old_size, write_length + 1, "%lld", number);
 }
 
+static VALUE obj_check_to_str(VALUE obj)
+{
+    VALUE str = rb_check_funcall(obj, id_to_str, 0, NULL);
+    if (RB_UNLIKELY(str != Qundef && !RB_TYPE_P(str, T_STRING))) {
+        rb_raise(rb_eTypeError, "%"PRIsVALUE"#to_str returned a non-String value of type %"PRIsVALUE,
+                rb_obj_class(obj), rb_obj_class(str));
+    }
+    return str;
+}
+
+static VALUE obj_to_s(VALUE obj)
+{
+    VALUE str = rb_funcall(obj, id_to_s, 0);
+    if (RB_TYPE_P(str, T_STRING))
+        return str;
+
+    VALUE to_str_result = obj_check_to_str(str);
```

Comment:
I'm curious as to why we try to call `to_str` on the return value of `to_s` if it does not return a string? Isn't it a bug in the caller if the object `to_s` does not return a string?

Reply:
This preserves the behaviour that ruby liquid has with `output << obj.to_s` where `to_s` is called explicitly, and `String#<<` calls `to_str` on the result.

An example I know about where this happens is `ActiveSupport::SafeBuffer`.  It's `#to_s` method returns itself, which seems reasonable given that `ActiveSupport::SafeBuffer` inherits from `String`.  However, it's `#to_str` (inherited from String#to_str) creates a String instance with its content duplicated.

Reply:
I think `String#<<` only calls `to_str` if `to_s` does not return a string-like value. For example:

This class `IsString` is string-like, so it does not call `to_str`:
```
irb(main):001:1* class IsString < String
irb(main):002:2*   def to_s
irb(main):003:2*     self
irb(main):004:1*   end
irb(main):005:2*   def to_str
irb(main):006:2*     "bar"
irb(main):007:1*   end
irb(main):008:0> end
=> :to_str
irb(main):009:0> f = IsString.new("foo")
irb(main):010:0> s = ''
irb(main):011:0> s << f
=> "foo"
```

But this class `IsNotString` is not string-like, so it does call `to_str`:
```
irb(main):001:1* class IsNotString
irb(main):002:2*   def to_s
irb(main):003:2*     self
irb(main):004:1*   end
irb(main):005:2*   def to_str
irb(main):006:2*     "bar"
irb(main):007:1*   end
irb(main):008:0> end
=> :to_str
irb(main):009:0> f = IsNotString.new
irb(main):010:0> s = ""
irb(main):011:0> s << f
=> "bar"
```

But I would argue that `IsNotString` has a bug, because `to_s` does not return a string.

Reply:
Oh, I see.  Ruby does a `RB_TYPE_P(s, T_STRING)` check and instances of a string derived class also have the T_STRING object type, so it can be used for ruby functions like `rb_str_append` that expected a RString struct and won't be converted by `StringValue`.

Thanks for clarifying that for me. I've pushed a commit to get rid of the to_str conversion that shouldn't be necessary.